### DISCUSSION TITLE
The pnpm i/up performance of pnpm monorepo (100+ modules) is poor. Is there any optimization suggestion or direction?

### DISCUSSION THREAD

**Discussion by rxliuli** [pnpm/pnpm#6552] [Q&A]
Currently we have 100+ modules, the number of modules in node_modules is more than 4k+, every time we use pnpm i or pnpm up it is around 40s, I want to know is there any way to speed up the process?

some thoughts

1. Configure `resolution-mode=time-based` -- I have tried, but the effect is not obvious, it may reduce about 5s
2. Exclude infrequently maintained business modules from pnpm-workspace.yaml to reduce the number of modules that pnpm cares about -- I haven't tried it yet, I want to know if anyone has tried it and is there any risk?
3. Merge different versions of the same dependency -- many dependencies will contain many different versions, which may be installed and introduced at different times

**Comment by literalpie** [comment]
I have a project with "380 workspace projects" reported when I run pnpm install. adding `--ignore-scripts` to `pnpm install` brings the time from 32s to 13s. (not much of a solution since scripts might be important, but maybe this is informative).

I tried profiling using `node --cpu-prof /Users/{{path-to-pnpm}}/pnpm install --ignore-scripts`. Here's a rough outline of the profile (I don't want to share the whole profile in case it includes something sensitive):

* First HeadlessInstall finishes around 2.4s
* SymlinkHoistedDependency happens a bunch of times until around 3.4s
* readFileAfterOpen happens a few times, but spread out enough that it takes about 0.2s, ending at 3.6
* tryReadProjectManifest happens a bunch, ending at 3.95
* readFileAfterClose until 4.2
* Another headlessInstall and symlinkDirectDependencies until 4.6s
* forceSymlink and headlessInstall ends at 5.9s
* **tryReadProjectManifest happens a ton of times until 7.3s**
* **readFileAfterClose happens a ton of times until 13.6s**
* writeCurrentLockfile is the last thing, ending around 14s

I'm really curious what it's doing during the bolded steps. If It's already resolved everything, and even logged out "already up to date" I don't understand why it seems to read every (or at least many) package.json file in the workspace. 

A similar test without `--ignore-scripts` seems to have another round of reading every package.json in the workspace

Trying with node-linker=hoisted takes a similar amount of time (14s with `--ignore-scripts`). It spends most of time on readPackageJson and tryReadProjectManifest
